### PR TITLE
MESH-1597: Split issuance and token creation

### PR DESCRIPTION
**Pull request by CJP10**
### new features

- `Asset::create_asset_and_mint` does what `Asset::create_asset` used to do

### modified API

- `Asset::create_asset` no longer mints tokens by default

### CODE REVIEW

File: pallets/asset/src/lib.rs [hunk 1]
```diff
@@ -1916,13 +1974,6 @@ impl<T: Trait> Module<T> {
             secondary_key,
         } = Identity::<T>::ensure_origin_call_permissions(origin)?;
 
-        // Check total supply here to avoid any later failure
-        Self::ensure_create_asset_parameters(&ticker, total_supply)?;
-        ensure!(
-            divisible || Self::is_unit_multiple(total_supply),
-            Error::<T>::InvalidTotalSupply
-        );
```
Comment (Centril) [inline pallets/asset/src/lib.rs:1924]: Looks like you're removing a bunch of checks in general from asset creation.
Also, the checks relevant to `total_supply` need to be moved to the `_and_mint` variant.

**Review by Centril — CHANGES_REQUESTED** [review]

File: pallets/asset/src/lib.rs [hunk 2]
```diff
@@ -521,6 +522,51 @@ decl_module! {
...
 
+        /// Initializes a new security token
+        /// makes the initiating account the owner of the security token
+        /// without minting the tokens. Use `issue` to mint the tokens.
+        ///
+        /// # Arguments
+        /// * `origin` - contains the secondary key of the caller (i.e. who signed the transaction to execute this function).
+        /// * `name` - the name of the token.
+        /// * `ticker` - the ticker symbol of the token.
+        /// * `total_supply` - the total supply of the token.
```
Comment (fmiguelgarcia) [inline pallets/asset/src/lib.rs:533]: doc error

File: pallets/asset/src/lib.rs [hunk 3]
```diff
@@ -521,6 +522,51 @@ decl_module! {
...
+        /// * `identifiers` - a vector of asset identifiers.
+        /// * `funding_round` - name of the funding round.
+        ///
+        /// ## Errors
+        /// - `InvalidAssetIdentifier` if any of `identifiers` are invalid.
+        /// - `MaxLengthOfAssetNameExceeded` if `name`'s length exceeds `T::AssetNameMaxLength`.
+        /// - `FundingRoundNameMaxLengthExceeded` if the name of the funding round is longer that
+        /// `T::FundingRoundNameMaxLength`.
+        /// - `AssetAlreadyCreated` if asset was already created.
+        /// - `TotalSupplyAboveLimit` if `total_supply > MAX_SUPPLY`.
```
Comment (fmiguelgarcia) [inline pallets/asset/src/lib.rs:545]: doc error

**Review by fmiguelgarcia — COMMENTED** [review]
Minor comments

File: pallets/asset/src/lib.rs [hunk 4]
```diff
@@ -521,6 +523,49 @@ decl_module! {
             Self::base_create_asset_and_mint(origin, name, ticker, total_supply, divisible, asset_type, identifiers, funding_round)
         }
 
+        /// Initializes a new security token
+        /// makes the initiating account the owner of the security token
+        /// without minting the tokens. Use `issue` to mint the tokens.
```
Comment (Centril) [inline pallets/asset/src/lib.rs:528]:
```suggestion
        /// Initializes a new security token, with the initiating account as its owner.
        /// The total supply will initially be zero. To mint tokens, use `issue`.
```

**Comment by Centril** [comment]
```
test_1            | Error: Extrinsic asset.createAsset expects 6 arguments, got 7.
test_1            |     at assert (/clicode/scripts/cli/node_modules/@polkadot/util/assert.js:34:11)
test_1            |     at extrinsicFn (/clicode/scripts/cli/node_modules/@polkadot/metadata/decorate/extrinsics/createUnchecked.js:23:22)
test_1            |     at Object.decorated [as createAsset] (/clicode/scripts/cli/node_modules/@polkadot/api/base/Decorate.js:346:46)
test_1            |     at Object.issueTokenPerDid (/clicode/scripts/cli/util/init.js:302:38)
test_1            |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
test_1            |     at async main (/clicode/scripts/cli/tests/4_permission_management.js:30:3)
```

Integration test needs to be fixed and I think this should be good to go.

File: pallets/asset/src/lib.rs [hunk 5]
```diff
@@ -503,12 +505,12 @@ decl_module! {
...
         /// ## Permissions
         /// * Portfolio
-        #[weight = <T as Trait>::WeightInfo::create_asset(
+        #[weight = <T as Trait>::WeightInfo::create_asset_and_mint(
             name.len() as u32,
             identifiers.len() as u32,
             funding_round.as_ref().map_or(0, |name| name.len()) as u32
         )]
-        pub fn create_asset(
+        pub fn create_asset_and_mint(
```
Comment (maxsam4) [inline pallets/asset/src/lib.rs:513]: This is not needed, let's remove it. Users can batch call `create_asset` and `issue` to get this functionality.

**Review by fmiguelgarcia — APPROVED** [review]

**Review by Centril — APPROVED** [review]

File: pallets/asset/src/lib.rs [hunk 6]
```diff
@@ -1019,13 +1014,34 @@ impl<T: Trait> AssetFnTrait<T::Balance, T::AccountId, T::Origin> for Module<T> {
...
+            name,
+            ticker,
+            divisible,
+            asset_type,
+            identifiers,
+            funding_round,
+        )
+    }
+
+    #[inline]
```
Comment (maxsam4) [inline pallets/asset/src/lib.rs:1033]: I don't think this should be inlined

**Review by maxsam4 — APPROVED** [review]

**Pull request merged**